src/cosalib/cmdlib.py

```diff
@@ -64,35 +66,66 @@ def run_verbose(args, **kwargs):
     return process
 
 
-def write_json(path, data):
+def get_lock_path(path):
+    """
+    Return the lock path to use for a given path.
+    """
+    dn = os.path.dirname(path)
+    bn = os.path.basename(path)
+    return os.path.join(dn, f".{bn}.lock")
+
+
+def write_json(path, data, lock_path=None):
     """
     Shortcut for writing a structure as json to the file system.
 
     :param path: The full path to the file to write
     :type: path: str
     :param data:  structure to write out as json
     :type data: dict or list
+    :param lock_path: path for the lock file to use
+    :type lock_path: string
     :raises: ValueError, OSError
     """
     dn = os.path.dirname(path)
     f = tempfile.NamedTemporaryFile(mode='w', dir=dn, delete=False)
     json.dump(data, f, indent=4)
     os.fchmod(f.file.fileno(), 0o644)
-    shutil.move(f.name, path)
+
+    # lock before moving
+    if not lock_path:
+        lock_path = get_lock_path(path)
+
+    with Lock(lock_path):
+        shutil.move(f.name, path)
 
 
-def load_json(path):
+def load_json(path, require_exclusive=True, lock_path=None):
     """
     Shortcut for loading json from a file path.
 
     :param path: The full path to the file
     :type: path: str
+    :param require_exclusive: lock file for exclusive read
+    :type require_exclusive: bool
+    :param lock_path: path for the lock file to use
+    :type lock_path: string
     :returns: loaded json
     :rtype: dict
     :raises: IOError, ValueError
     """
-    with open(path) as f:
-        return json.load(f)
+    lock = None
+    if require_exclusive:
+        if not lock_path:
+            lock_path = get_lock_path(path)
+        lock = Lock(lock_path)
+        lock.lock()
+    try:
+        with open(path) as f:
+            return json.load(f)
+    finally:
+        if lock:
+            lock.unlock(unconditionally=True)
```

Comment:
I'm not sure if I follow this. If we need exclusive access to `meta.json`, then the lock should outlive this function call, right? I.e. we need e.g. `cosa buildextend-openstack` to be able to:
1. get exclusive lock
2. do merge
3. write back
4. release lock

As @cgwalters mentioned, locking the read and write operations themselves is not necessary, because we already use `rename()` to update `meta.json`.

It seems like what we want is a new library function call, e.g. `merge_json` which takes the path to the `meta.json` and a list of `meta.$artifact.json`, which then does the steps above. Then `cosa meta-merge` would just call out to that.

Reply:
Who knows if this has any design sense, but my thinking was to add a callback func to `write_json`, e.g. `write_json(data, path, merge_func=....)`  where `write_json` would call the merge_func with the on-disk and data version to maintain the lock. 

e.g.  `cmdlib.meta.write` would use:
      `write_json(self._meta.path, dict(self), merge_func=meh)`

And in `write_json` would have something like:
```
￼    with Lock(lock_path):
         if merge_func:
              with open(path, 'r') as f:
                  on_disk = json.loads(f.read()))
             data = merge_func(on_disk, data.copy())

        with open(path, 'w') as f:
              f.write(json.dumps(data, indent=4))
```

Part of why I wanted to do this in `cmdlib.py` was that it opens the ability work on `builds.json` too; we could have parallel builds, as well as parallel build extends.

Reply:
Gotcha, totally valid approach IMO!

Maybe let's mention that in the commit message. E.g.:

> This is prep for adding callbacks to `write_json` for doing more complex operations under the lock.

Do we need locking on the read side, though?

Also, this doesn't yet implement the enhancement, right? Let's drop that commit message footer?

Reply:
updated the commit message.

> Do we need locking on the read side, though?

Maybe. I didn't see any downsides to having locking on the read operation _and_ that's why I made it optional. It doesn't hurt.